storage/storage-interface/src/sharded_state_update_refs.rs

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_types::{
+    state_store::{state_key::StateKey, state_value::StateValue, NUM_STATE_SHARDS},
+    write_set::WriteSet,
+};
+use arr_macro::arr;
+
+pub type StateUpdateRefWithOffset<'kv> = (usize, &'kv StateKey, Option<&'kv StateValue>);
+
+pub struct ShardedStateUpdateRefs<'kv> {
+    /// (idx, key, value) tuple per update, sharded
+    ///    Box<[]> to release over allocated memory during construction
+    pub shards: [Box<[StateUpdateRefWithOffset<'kv>]>; NUM_STATE_SHARDS],
+    pub num_versions: usize,
+}
+
+impl<'kv> ShardedStateUpdateRefs<'kv> {
+    pub fn index_write_sets(
+        write_sets: impl IntoIterator<Item = &'kv WriteSet>,
+        num_write_sets: usize,
+    ) -> Self {
+        Self::index_per_version_updates(
+            write_sets
+                .into_iter()
+                .map(|write_set| write_set.state_update_refs()),
+            num_write_sets,
+        )
+    }
+
+    pub fn index_per_version_updates<
+        UpdateIter: IntoIterator<Item = (&'kv StateKey, Option<&'kv StateValue>)>,
+        VersionIter: IntoIterator<Item = UpdateIter>,
+    >(
+        updates_by_version: VersionIter,
+        num_versions: usize,
+    ) -> Self {
+        let mut shards = arr![Vec::with_capacity(num_versions / 16); 16];
```

Comment:
why / 16?

Reply:
don't want to over allocate too much. As long as it's not empty, we shouldn't hit too many reallocates, right?

Reply:
I'd say at least `/ 8` since most of txns touches at least 2 states.